autoload/ale/engine.vim

```diff
@@ -0,0 +1,218 @@
+        return
+    endif
+
+    let job_info = s:job_info_map[job_id]
+
+    call ale#engine#ClearJob(a:job)
+
+    let linter = job_info.linter
+    let output = job_info.output
+    let buffer = job_info.buffer
+
+    let linter_loclist = ale#util#GetFunction(linter.callback)(buffer, output)
+
+    " Make some adjustments to the loclists to fix common problems.
+    call ale#util#FixLocList(buffer, linter_loclist)
+
+    if g:ale_buffer_should_reset_map[buffer]
+        let g:ale_buffer_should_reset_map[buffer] = 0
+        let g:ale_buffer_loclist_map[buffer] = []
+    endif
+
+    " Add the loclist items from the linter.
+    call extend(g:ale_buffer_loclist_map[buffer], linter_loclist)
+
+    " Sort the loclist again.
+    " We need a sorted list so we can run a binary search against it
+    " for efficient lookup of the messages in the cursor handler.
+    call sort(g:ale_buffer_loclist_map[buffer], 'ale#util#LocItemCompare')
+
+    if g:ale_set_loclist
+        call setloclist(0, g:ale_buffer_loclist_map[buffer])
+    endif
+
+    if g:ale_set_signs
+        call ale#sign#SetSigns(buffer, g:ale_buffer_loclist_map[buffer])
+    endif
+
+    " Mark line 200, column 17 with a squiggly line or something
+    " matchadd('ALEError', '\%200l\%17v')
+endfunction
+
+function! ale#engine#HandleExitNeoVim(job, data, event) abort
+    call s:HandleExit(a:job)
+endfunction
+
+function! ale#engine#HandleExitVim(channel) abort
+    call s:HandleExit(ch_getjob(a:channel))
+endfunction
+
+function! ale#engine#ApplyLinter(buffer, linter) abort
+    if has_key(a:linter, 'job')
+        " Stop previous jobs for the same linter.
+        call ale#engine#ClearJob(a:linter.job)
+    endif
+
+    if has_key(a:linter, 'command_callback')
+        " If there is a callback for generating a command, call that instead.
+        let command = ale#util#GetFunction(a:linter.command_callback)(a:buffer)
+    else
+        let command = a:linter.command
+    endif
+
+    if command =~# '%s'
+        " If there is a '%s' in the command string, replace it with the name
+        " of the file.
+        let command = printf(command, shellescape(fnamemodify(bufname(a:buffer), ':p')))
+    endif
+
+    if has('nvim')
+        if a:linter.output_stream ==# 'stderr'
+            " Read from stderr instead of stdout.
+            let job = jobstart(command, {
+            \   'on_stderr': 'ale#engine#GatherOutputNeoVim',
+            \   'on_exit': 'ale#engine#HandleExitNeoVim',
+            \})
+        elseif a:linter.output_stream ==# 'both'
+            let job = jobstart(command, {
+            \   'on_stdout': 'ale#engine#GatherOutputNeoVim',
+            \   'on_stderr': 'ale#engine#GatherOutputNeoVim',
+            \   'on_exit': 'ale#engine#HandleExitNeoVim',
+            \})
+        else
+            let job = jobstart(command, {
+            \   'on_stdout': 'ale#engine#GatherOutputNeoVim',
+            \   'on_exit': 'ale#engine#HandleExitNeoVim',
```

Comment:
Could these not still be in script functions, so they cannot be accessed from the outside? If so, that would provide better encapsulation.